### ISSUE TITLE
Broken build in OS X 10.10

### ISSUE THREAD

**Issue by zanchey** [fish-shell/fish-shell#4029]
2.6b1 [failed to build on OS X 10.10 Yosemite](https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/1340/version=yosemite/testReport/junit/brew-test-bot/yosemite/install___devel_fish/):

```
  CXX      obj/builtin_commandline.o
In file included from src/autoload.cpp:11:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:602:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/utility:259:9: error: field has incomplete type 'lru_cache_t<autoload_t, autoload_function_t>::lru_node_t'
    _T2 second;
        ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/map:588:16: note: in instantiation of template class 'std::__1::pair<const std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>' requested here
    value_type __cc;
               ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__tree:595:16: note: in instantiation of template class 'std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>' requested here
    value_type __value_;
               ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__tree:617:22: note: in instantiation of template class 'std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, void *>' requested here
    typedef typename __node::base                                 __node_base;
                     ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/map:650:19: note: in instantiation of template class 'std::__1::__tree_iterator<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, void *> *, long>' requested here
    _TreeIterator __i_;
                  ^
./src/lru.h:42:21: note: in instantiation of template class 'std::__1::__map_iterator<std::__1::__tree_iterator<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, void *> *, long> >' requested here
        node_iter_t iter;
                    ^
./src/lru.h:105:20: note: in instantiation of member class 'lru_cache_t<autoload_t, autoload_function_t>::lru_node_t' requested here
        evict_node(static_cast<lru_node_t *>(mouth.prev));
                   ^
./src/lru.h:271:13: note: in instantiation of member function 'lru_cache_t<autoload_t, autoload_function_t>::evict_last_node' requested here
            evict_last_node();
            ^
./src/autoload.h:63:41: note: in instantiation of member function 'lru_cache_t<autoload_t, autoload_function_t>::evict_all_nodes' requested here
    void remove_all_functions() { this->evict_all_nodes(); }
                                        ^
./src/lru.h:24:12: note: definition of 'lru_cache_t<autoload_t, autoload_function_t>::lru_node_t' is not complete until the closing '}'
    struct lru_node_t;
           ^
In file included from src/autoload.cpp:11:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:602:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/utility:259:9: error: field has incomplete type 'lru_cache_t<autoload_t, autoload_function_t>::lru_node_t'
    _T2 second;
        ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/map:589:21: note: in instantiation of template class 'std::__1::pair<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>' requested here
    __nc_value_type __nc;
                    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__tree:595:16: note: in instantiation of template class 'std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>' requested here
    value_type __value_;
               ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__tree:617:22: note: in instantiation of template class 'std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, void *>' requested here
    typedef typename __node::base                                 __node_base;
                     ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/map:650:19: note: in instantiation of template class 'std::__1::__tree_iterator<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, void *> *, long>' requested here
    _TreeIterator __i_;
                  ^
./src/lru.h:42:21: note: in instantiation of template class 'std::__1::__map_iterator<std::__1::__tree_iterator<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, void *> *, long> >' requested here
        node_iter_t iter;
                    ^
./src/lru.h:105:20: note: in instantiation of member class 'lru_cache_t<autoload_t, autoload_function_t>::lru_node_t' requested here
        evict_node(static_cast<lru_node_t *>(mouth.prev));
                   ^
./src/lru.h:271:13: note: in instantiation of member function 'lru_cache_t<autoload_t, autoload_function_t>::evict_last_node' requested here
            evict_last_node();
            ^
./src/autoload.h:63:41: note: in instantiation of member function 'lru_cache_t<autoload_t, autoload_function_t>::evict_all_nodes' requested here
    void remove_all_functions() { this->evict_all_nodes(); }
                                        ^
./src/lru.h:24:12: note: definition of 'lru_cache_t<autoload_t, autoload_function_t>::lru_node_t' is not complete until the closing '}'
    struct lru_node_t;
           ^
./src/lru.h:105:20: error: static_cast from 'lru_cache_t<autoload_t, autoload_function_t>::lru_link_t *' to 'lru_cache_t<autoload_t, autoload_function_t>::lru_node_t *', which are not related by inheritance, is not allowed
        evict_node(static_cast<lru_node_t *>(mouth.prev));
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./src/lru.h:271:13: note: in instantiation of member function 'lru_cache_t<autoload_t, autoload_function_t>::evict_last_node' requested here
            evict_last_node();
            ^
./src/autoload.h:63:41: note: in instantiation of member function 'lru_cache_t<autoload_t, autoload_function_t>::evict_all_nodes' requested here
    void remove_all_functions() { this->evict_all_nodes(); }
                                        ^
In file included from src/autoload.cpp:18:
In file included from ./src/autoload.h:11:
In file included from ./src/lru.h:7:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/map:974:44: error: no viable conversion from 'iterator' (aka '__tree_iterator<value_type, __node_pointer, difference_type>') to 'iterator' (aka '__map_iterator<typename __base::iterator>')
          iterator end() _NOEXCEPT {return __tree_.end();}
                                           ^~~~~~~~~~~~~
./src/lru.h:199:37: note: in instantiation of member function 'std::__1::map<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t, std::__1::less<std::__1::basic_string<wchar_t> >, std::__1::allocator<std::__1::pair<const std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t> > >::end' requested here
        if (where == this->node_map.end()) return false;
                                    ^
src/autoload.cpp:65:60: note: in instantiation of member function 'lru_cache_t<autoload_t, autoload_function_t>::evict_node' requested here
int autoload_t::unload(const wcstring &cmd) { return this->evict_node(cmd); }
                                                           ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/map:648:29: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'iterator' (aka '__tree_iterator<value_type, __node_pointer, difference_type>') to 'const std::__1::__map_iterator<std::__1::__tree_iterator<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, void *> *, long> > &' for 1st argument
class _LIBCPP_TYPE_VIS_ONLY __map_iterator
                            ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/map:648:29: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'iterator' (aka '__tree_iterator<value_type, __node_pointer, difference_type>') to 'std::__1::__map_iterator<std::__1::__tree_iterator<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, void *> *, long> > &&' for 1st argument
class _LIBCPP_TYPE_VIS_ONLY __map_iterator
                            ^
In file included from src/autoload.cpp:18:
In file included from ./src/autoload.h:11:
./src/lru.h:188:13: error: no viable conversion from 'std::__1::__map_iterator<std::__1::__tree_iterator<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, void *> *, long> >' to 'const std::__1::__map_const_iterator<std::__1::__tree_const_iterator<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, void *> *, long> >'
        if (where == this->node_map.end()) {
            ^~~~~
src/autoload.cpp:125:39: note: in instantiation of member function 'lru_cache_t<autoload_t, autoload_function_t>::get' requested here
    autoload_function_t *func = this->get(cmd);
                                      ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/map:739:5: note: candidate constructor
    __map_const_iterator(
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/map:713:29: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'std::__1::__map_iterator<std::__1::__tree_iterator<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, void *> *, long> >' to 'const std::__1::__map_const_iterator<std::__1::__tree_const_iterator<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, void *> *, long> > &' for 1st argument
class _LIBCPP_TYPE_VIS_ONLY __map_const_iterator
                            ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/map:713:29: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'std::__1::__map_iterator<std::__1::__tree_iterator<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, void *> *, long> >' to 'std::__1::__map_const_iterator<std::__1::__tree_const_iterator<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, void *> *, long> > &&' for 1st argument
class _LIBCPP_TYPE_VIS_ONLY __map_const_iterator
                            ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/map:737:5: note: candidate constructor not viable: no known conversion from 'std::__1::__map_iterator<std::__1::__tree_iterator<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, void *> *, long> >' to 'std::__1::__tree_const_iterator<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<wchar_t>, lru_cache_t<autoload_t, autoload_function_t>::lru_node_t>, void *> *, long>' for 1st argument
    __map_const_iterator(_TreeIterator __i) _NOEXCEPT : __i_(__i) {}
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/map:770:49: note: passing argument to parameter '__x' here
    bool operator==(const __map_const_iterator& __x, const __map_const_iterator& __y)
                                                ^
In file included from src/autoload.cpp:18:
In file included from ./src/autoload.h:11:
./src/lru.h:66:21: error: comparison of distinct pointer types ('lru_cache_t<autoload_t, autoload_function_t>::lru_node_t *' and 'lru_cache_t<autoload_t, autoload_function_t>::lru_link_t *')
        assert(node != &mouth);
               ~~~~ ^  ~~~~~~
./src/common.h:218:20: note: expanded from macro 'assert'
#define assert(e) (e) ? ((void)0) : __assert(#e, __FILE__, __LINE__, 0)
                   ^
./src/lru.h:192:9: note: in instantiation of member function 'lru_cache_t<autoload_t, autoload_function_t>::promote_node' requested here
        promote_node(&where->second);
        ^
src/autoload.cpp:125:39: note: in instantiation of member function 'lru_cache_t<autoload_t, autoload_function_t>::get' requested here
    autoload_function_t *func = this->get(cmd);
                                      ^
In file included from src/autoload.cpp:18:
In file included from ./src/autoload.h:11:
./src/lru.h:74:26: error: assigning to 'lru_cache_t<autoload_t, autoload_function_t>::lru_link_t *' from incompatible type 'lru_cache_t<autoload_t, autoload_function_t>::lru_node_t *'
        node->next->prev = node;
                         ^ ~~~~
./src/lru.h:76:20: error: assigning to 'lru_cache_t<autoload_t, autoload_function_t>::lru_link_t *' from incompatible type 'lru_cache_t<autoload_t, autoload_function_t>::lru_node_t *'
        mouth.next = node;
                   ^ ~~~~
8 errors generated.
```

**Comment by ridiculousfish** [comment]
Nice find, I'll take a whack at it.

**Comment by krader1961** [comment]
I don't have a OS X Yosemite VM setup but can do so if it is needed.

**Comment by ridiculousfish** [comment]
Fixed as 523dc6da6da263629f04848e8c87267b9444e3d5
